feat(c11): Adapt code to be compatible with ISO C11

miguelafsilva5 opened the pull request:
## PR Description

This PR is the first on the chain of PRs (This PR <- #146  <- #147 ). 
The main goal of this PR is to force ISO C11 on bao and deal with all -pedantic errors. Each commit tries to deal with a specific error. 



The platforms tested were:
- qemu-aarch64-virt
- qemo-riscv64-virt

josecm commented:
I don't think we need the wip in #120. We can use just this PR for both enabling pedantic C11 and adding the fixes to the new errors.

josecm commented:
@miguelafsilva5 please fix your commit messages to pass gitlint checks and apply the code formatted.

I also see that the builds are still failing due to `-Wpedantic` errors.

josecm commented on src/arch/armv8/armv8-a/aarch64/vmm.c:
```diff
@@ -18,7 +18,7 @@ void vmm_arch_init_tcr()
      */
 
-    static size_t min_parange = 0b111;
+    static size_t min_parange = 0x7;//0b111;
```
We could leave the comments with the bit pattern ou. If we are keeping these comments at least clean them up:

```suggestion
    static size_t min_parange = 0x7; // 0b111
```

miguelafsilva5 replied:
Fixed in commit a27a15bff8ad78e2cabf764fe795b62650daf18f

josecm commented on src/arch/armv8/vgicv3.c:
```diff
@@ -129,7 +129,8 @@ void vgicd_emul_router_access(struct emul_access* acc, struct vgic_reg_handler_i
 
     if (interrupt == NULL) {
-        return vgic_emul_razwi(acc, handlers, gicr_access, vgicr_id);
+        vgic_emul_razwi(acc, handlers, gicr_access, vgicr_id); //Check again
```
Don't get what the comment is there for.

miguelafsilva5 replied:
Fixed in commit 334634b63fc85d1f5e110a32f46d95ab8ddea2a5

josecm commented on src/arch/riscv/inc/arch/csrs.h:
```diff
@@ -217,7 +217,7 @@
 #ifndef __ASSEMBLER__
 
-#define CSRR(csr)                                                            \
+/*#define CSRR(csr)                                                            \
```
Yes. Remove.

josecm commented on src/lib/inc/util.h:
```diff
@@ -79,9 +79,12 @@ static inline bool range_in_range(unsigned long base1, unsigned long size1, unsi
+#define __DEFINED(VALUE)                ___DEFINED(VALUE true, false, dummy)
 #define ___DEFINED(IGNORE, RESULT, ...) (RESULT)
 
+#define EMPTY_STRUCT_FIELDS \
+    uint8_t dummy_;
+
```
This macro is a good idea! :+1:

josecm commented on src/lib/inc/util.h:
```diff
@@ -79,7 +79,7 @@ static inline bool range_in_range(unsigned long base1, unsigned long size1, unsi
 #define _DEFINED_1                      0,
 #define _DEFINED(VALUE)                 __DEFINED(_DEFINED_##VALUE)
-#define __DEFINED(VALUE)                ___DEFINED(VALUE true, false)
+#define __DEFINED(VALUE)                ___DEFINED(VALUE true, false, dummy)
```
@DavidMCerdeira Yes. ISO C does not allow empty ellipses arguments which happens in one of the cases of this macro chain. But it would be a good idea to clarify the original error in the commit message.

DavidMCerdeira replied:
Yes I agree

miguelafsilva5 replied:
Fixed in commit 6c69e52d29556667ce66240bbe0306974edf05cb

miguelafsilva5 replied:
Fixed in commit 6c69e52d29556667ce66240bbe0306974edf05cb

josecm commented on src/core/inc/bao.h:
```diff
@@ -14,15 +14,26 @@
 
-#define WARNING(args, ...) console_printk("BAO WARNING: " args "\n" __VA_OPT__(, ) __VA_ARGS__);
+#define INFO(...)                                         \
+    do {                                                   \
+        console_printk("BAO INFO: " __VA_ARGS__); \
+        while (true) { }                                   \
+    } while (true);
```
I think what you want here is a single iteration while loop so it would be better to:

```suggestion
    do {                                                   \
        console_printk("BAO INFO: " __VA_ARGS__); \
        while (true) { }                                   \
    } while (false)
```

josecm replied:
You also frame the commit message from this commit regarding only the VA_OPT issue. No mention why you introduce the `do { ... } while()` loops.

miguelafsilva5 replied:
Fixed in commit e6bc02bdc72f79dcdb517828a42d7e57842c3ed5

josecm reviewed (changes requested):
Finally, the scopes you are using in your commit messages don't make much sense. Scopes are supposed to be more of a component of subsystem of the code or a topic regarding the modification, not a keyword for the modification you are introducing. As many of these are related to ISO C, C11, pedantic c, I would suggest using some scope in this sense or none at all.

miguelafsilva5 commented on Makefile:
```diff
@@ -235,7 +235,7 @@ $(build_dir)/%.d : $(src_dir)/%.[c,S]
 $(objs-y):
 	@echo "Compiling source	$(patsubst $(cur_dir)/%, %, $<)"
-	-@$(cc) $(CFLAGS) -c $< -o $@
+	@$(cc) $(CFLAGS) -c $< -o $@
```
I was supressing errors while working. It should have never reached the PR. It is fixed already.

danielRep reviewed (commented):
The changes seem OK to me and well-documented on each commit description. However, build checkers are failing and, at least locally, the format-check for me is also failing. I don't understand how the checker didn't fail too. 

```
13:56 $ make format-check 
18103,18105c18103,18105
< #define CPU_MSG_HANDLER(handler, handler_id)                \
<     __attribute__((section(".ipi_cpumsg_handlers"), used))  \
<     cpu_msg_handler_t __cpumsg_handler_##handler = handler; \
---
> #define CPU_MSG_HANDLER(handler, handler_id)                           \
>     __attribute__((section(".ipi_cpumsg_handlers"),                    \
>         used)) cpu_msg_handler_t __cpumsg_handler_##handler = handler; \
make: *** [/home/daniel/workspace/osyxtech/bao-hypervisor/ci/ci.mk:147: format-check] Error 1
```

miguelafsilva5 commented on src/arch/riscv/inc/arch/csrs.h:
```diff
@@ -227,6 +227,56 @@
 #define CSRC(csr, rs) __asm__ volatile("csrc  " XSTR(csr) ", %0\n\r" ::"rK"(rs) : "memory")
+*/
+
+#define CSRS_GEN_ACCESSORS_NAMED(csr_name, csr_id) \
```
Fixed in commit f5b9d651b991c97dcb7d3b6706296fefd0d752b9

miguelafsilva5 commented on src/arch/armv8/vgic.c:
```diff
@@ -693,7 +693,7 @@ void vgic_emul_generic_access(struct emul_access* acc, struct vgic_reg_handler_i
 struct vgic_reg_handler_info isenabler_info = {
     vgic_emul_generic_access,
-    0b0100,
+    0x4,//0b0100,
```
Fixed in commit a27a15bff8ad78e2cabf764fe795b62650daf18f

miguelafsilva5 commented on src/core/inc/bao.h:
```diff
@@ -14,15 +14,26 @@
+    do {                                                   \
+        console_printk("BAO WARNING: " __VA_ARGS__); \
+        while (true) { }                                   \
+    } while (true);
```
Fixed in commit e6bc02bdc72f79dcdb517828a42d7e57842c3ed5

miguelafsilva5 commented:
> The changes seem OK to me and well-documented on each commit description. However, build checkers are failing.

Commit 1853b59efa741a0a1f622155c5fb528e71145c52 seems to have fixed it.

DavidMCerdeira reviewed (commented):
Commit message for commit f2a7e5008b23547c33b9eae98cd2a700e149cccf has extra text before the sign-off

DavidMCerdeira reviewed (commented):
In e6bc02bdc72f79dcdb517828a42d7e57842c3ed5, nit: "Its expansion ends in a variadic macro that was not receive the correct amount of arguments if the evaluated macro was not define." should be: "Its expansion
ends in a variadic macro that was not receiv**ing** the correct **number** of arguments
if the evaluated macro was not define**d**."

miguelafsilva5 commented:
> Commit message for commit [f2a7e50](https://github.com/bao-project/bao-hypervisor/commit/f2a7e5008b23547c33b9eae98cd2a700e149cccf) has extra text before the sign-off

Fixed in bd1b162fb26d7a29c2ecd1b95d791788d510a80e

> In commit message for [e6bc02b](https://github.com/bao-project/bao-hypervisor/commit/e6bc02bdc72f79dcdb517828a42d7e57842c3ed5) should it be **VA_OTPS** or **VA_OPT** ?

Fixed in c4edbd3b53be0894531c4a267882304a8098038a

> In [e6bc02b](https://github.com/bao-project/bao-hypervisor/commit/e6bc02bdc72f79dcdb517828a42d7e57842c3ed5), nit: "Its expansion ends in a variadic macro that was not receive the correct amount of arguments if the evaluated macro was not define." should be: "Its expansion ends in a variadic macro that was not receiv**ing** the correct **number** of arguments if the evaluated macro was not define**d**."

@DavidMCerdeira you tagged the wrong commit, but it is fixed here 0f83bbf091474cb6dffe0e44a8421820a784079d

DavidMCerdeira reviewed (changes requested):
There are compilations errors for tx2, ultra96, zcu102, zcu104 platforms
These are related to the uart drivers

miguelafsilva5 commented:
> There are compilations errors for tx2, ultra96, zcu102, zcu104 platforms These are related to the uart drivers

The errors were related to previous commits. I've fixed them in commits ee8c6b474e5463571e168e7b2c58aa77b7381916 and 88d90e9eeee727708f4a05f8d12662a57e338f9e
